### PR TITLE
Fix 22306 web appropriate message is not shown when drag drop a folder

### PR DESCRIPTION
**Pull request by Charan-hs**
### Fixed Issues

$ https://github.com/Expensify/App/issues/22306
PROPOSAL: https://github.com/Expensify/App/issues/22306#issuecomment-1631437724

### Tests
1. Open Any Chat.
2. Drag and Drop a Folder on the Chat window.
3. An Error message of `Uploading a folder is not allowed. Try a different file.` shown.

 

- [x] Verify that no errors appear in the JS console

### Offline tests
same as above

### QA Steps
same as above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

1. Safari 

https://github.com/Expensify/App/assets/62904657/cb7a49f1-e22f-461b-9f26-e9fd4eff31cc

2. Chrome

https://github.com/Expensify/App/assets/62904657/874c710a-ca50-4d8b-a843-4aa47d2426fb

3. Uploading file without type property 

https://github.com/Expensify/App/assets/62904657/886c883f-a341-4b60-ad93-c61baa03f816

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/62904657/38babcc4-d5c2-4aa0-b180-dc2d2e6f9667

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/62904657/054e49ef-b09e-44c7-97a9-3ab03ad27aad

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/62904657/69302d96-3982-49a6-9f4c-e78e2152f16e

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/62904657/a84342ac-813c-468f-b016-bfdf7cf90fc1
</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/62904657/40808c3f-9e27-4896-b442-d75260309622

</details>

### CODE REVIEW

**Comment by Charan-hs** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Charan-hs** [comment]
recheck

**Comment by Charan-hs** [comment]
@abdulrahuman5196  we can modify the Error Title and message Based on your need.

File: src/components/AttachmentModal.js
Comment (abdulrahuman5196) [inline src/components/AttachmentModal.js:185]: Why do we need both file and data? `_data.getAsFile` provides the file right?

Comment (Charan-hs) [reply src/components/AttachmentModal.js:185]: yeah, but this function is also used in the file picker addition to drag and drop, so there you won't get data files but normal file objects. Added this condition will resolve even if a user is uploading on different platforms it won't have an effect. 
we can remove `_data` props but we should add this check
```js
let _file = _data
if (typeof _data.getAsFile === 'function' ){
     _file = _data.getAsFile();
 }
```
so what's your suggestion?

Comment (Charan-hs) [reply src/components/AttachmentModal.js:185]: @abdulrahuman5196  

```js
 /**
     * @param {Object} _data
     * @returns {Object}
     */
    const isDirectoryCheckAndGetFile = useCallback((_data) => {
        if (typeof _data.webkitGetAsEntry === 'function' && _data.webkitGetAsEntry().isDirectory) {
            setIsAttachmentInvalid(true);
            setAttachmentInvalidReasonTitle(props.translate('attachmentPicker.attachmentError'));
            setAttachmentInvalidReason(props.translate('attachmentPicker.folderNotAllowedMessage'));
            return null;
        }
        if (typeof _data.getAsFile === 'function' ){
            return _data.getAsFile();
        }
        return _data

    },[props.translate])

const validateAndDisplayFileToUpload = ...(_data) => {
const _file = isDirectoryCheckAndGetFile(_data)

if(!_file){
return }
....

}...
```

what about this approach?

Comment (abdulrahuman5196) [reply src/components/AttachmentModal.js:185]: IMO, IsDirectorycheck and getFile is different functionality. So IsDirectorycheck is fine and getFile can stay in current validateAndDisplayFileToUpload place

Comment (Charan-hs) [reply src/components/AttachmentModal.js]: done

**Comment by abdulrahuman5196** [comment]
recheck

**Comment by abdulrahuman5196** [comment]
@luacmartins If possible could you start the workflow steps to avoid lint failures after testing?

**Comment by luacmartins** [comment]
done

File: src/languages/en.js
Comment (abdulrahuman5196) [inline src/languages/en.js:171]: @luacmartins Sorry to bother again.
Whom should we confirm the error message with? In most cases previously internal engineers would provide the required string AFAIK.

**Comment by abdulrahuman5196** [comment]
@Charan-hs lint and signed commits are failing, kindly fix the same(I suspected lint to fail, that's why requested to run checks).

Since you are a first time contributor just providing some information - `PR Reviewer Checklist` would be fixed after my review. Other checks should be fixed before that.

**Comment by Charan-hs** [comment]
@abdulrahuman5196 

all Commits are Signed and resolved lint issues.

**Comment by Charan-hs** [comment]
recheck

File: src/components/AttachmentModal.js
Comment (abdulrahuman5196) [inline src/components/AttachmentModal.js:238]: Why do we need to want to disable a lint rule here? Can we just rename the variable @Charan-hs ?

File: src/languages/en.js
Comment (abdulrahuman5196) [reply src/languages/en.js]: @luacmartins Gentle ping on this confirmation

File: src/components/AttachmentModal.js
Comment (Charan-hs) [reply src/components/AttachmentModal.js:238]: Just to stay with the previous file name so. can I change that to `fileObject` @abdulrahuman5196  ?

File: src/languages/en.js
Comment (luacmartins) [reply src/languages/en.js]: @abdulrahuman5196 thanks for the ping, I missed the first one 😬 I added the `Waiting for copy` label in the issue to get someone to review the copy. Let's see what they say.

Comment (luacmartins) [reply src/languages/en.js]: @Charan-hs @abdulrahuman5196 the updated copy is: `Uploading a folder is not allowed. Try a different file.`

File: src/languages/en.js
Comment (luacmartins) [inline src/languages/en.js:171]:
```suggestion
        folderNotAllowedMessage: 'Uploading a folder is not allowed. Try a different file.',
```

File: src/languages/es.js
Comment (luacmartins) [inline src/languages/es.js:170]: This translation seems off to me. Did you request it to be translated, e.g. https://expensify.slack.com/archives/C01GTK53T8Q/p1641913188292300

Comment (Charan-hs) [reply src/languages/es.js:170]: Requested Now.

Comment (Charan-hs) [reply src/languages/es.js]: Got translation from @rocio https://expensify.slack.com/archives/C01GTK53T8Q/p1690482354912629

**Comment by Charan-hs** [comment]
@abdulrahuman5196  @luacmartins  resolved merge conflicts

**Review by luacmartins — COMMENTED** [review]
Looking good. Just a couple small comments.

File: src/components/AttachmentModal.js
Comment (Charan-hs) [reply src/components/AttachmentModal.js:217]: @luacmartins Based on this https://github.com/Expensify/App/pull/23321#discussion_r1271442232  discussion with @abdulrahuman5196, we came to this approach. open for any suggestions.

File: src/components/AttachmentModal.js
Comment (Charan-hs) [reply src/components/AttachmentModal.js:227]: This type of Approach is used in the entire file so used the same. we can remove this disabled Lint rule by adding `props` as a dependency rather than `props.translate`.

- Warning message from the `eslint` checker

`React Hook useCallback has a missing dependency: 'props'. Either include it or remove the dependency array. However, 'props' will change when *any* prop changes, so the preferred fix is to destructure the 'props' object outside of the useCallback call and refer to those specific props inside useCallback`

File: src/components/AttachmentModal.js
Comment (luacmartins) [reply src/components/AttachmentModal.js:217]: Sounds good! Thanks for the link!

File: src/components/AttachmentModal.js
Comment (luacmartins) [reply src/components/AttachmentModal.js:227]: Can we use the `useLocalize` hook instead of getting the translate prop from Onyx and update those instances, like we do [here](https://github.com/Expensify/App/blob/0f40047b279ff5ddcdab798d13d380d555789351/src/components/CountryPicker/CountrySelectorModal.js#L46)?

Comment (Charan-hs) [reply src/components/AttachmentModal.js:227]: @luacmartins , can I update Entrie `AttachModal.js` from `props.translate` to `translate` from `useLocalize`?

Comment (luacmartins) [reply src/components/AttachmentModal.js:227]: Yea, I think that should be fine

Comment (Charan-hs) [reply src/components/AttachmentModal.js]: Done with changes

**Review by luacmartins — DISMISSED** [review]
LGTM all yours @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
@Charan-hs Can we also check and fix the merge conflicts?

**Comment by abdulrahuman5196** [comment]
@Charan-hs Is the changes complete from your end?

**Comment by abdulrahuman5196** [comment]
@luacmartins I already tested previously all the changes before merge conflict and it was working fine. Can we re-run the workflow checks? If its working I can do a quick test again and we can merge.

**Comment by luacmartins** [comment]
@abdulrahuman5196 kicked off the workflow checks

**Comment by abdulrahuman5196** [comment]
@Charan-hs Lint code is failing. Could you kindly check on the same?

**Comment by luacmartins** [comment]
@Charan-hs lint is failing!

**Comment by luacmartins** [comment]
All checks are passing now! @abdulrahuman5196 wanna do a final check before we merge?

**Comment by abdulrahuman5196** [comment]
Yes. Working on it

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>





https://github.com/Expensify/App/assets/46707890/f4048c3d-b0fb-4496-a3b5-d478dc40b289


</details>

<details>
<summary>Mobile Web - Chrome</summary>





https://github.com/Expensify/App/assets/46707890/05b65d76-a810-47e0-8d80-66ae56b1e7e4


</details>

<details>
<summary>Mobile Web - Safari</summary>








https://github.com/Expensify/App/assets/46707890/8e65efba-86bc-4464-ac2c-7a9080a14791


</details>

<details>
<summary>Desktop</summary>






https://github.com/Expensify/App/assets/46707890/728d6762-7787-4271-a14c-9b43b196a7f9



</details>

<details>
<summary>iOS</summary>







https://github.com/Expensify/App/assets/46707890/ca192e76-52d7-4cc1-9e6e-5e482f5c6663

</details>

<details>
<summary>Android</summary>





https://github.com/Expensify/App/assets/46707890/fe62e42c-0c4d-4dd6-8f97-5b0ac26da1cb


https://github.com/Expensify/App/assets/46707890/758b1e68-5d65-400b-8ea0-8079c3d78c4a


</details>

**Review by abdulrahuman5196 — APPROVED** [review]
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @luacmartins Good to merge

🎀 👀 🎀
C+ Reviewed

**Review by luacmartins — APPROVED** [review]

**Comment by abdulrahuman5196** [comment]
Just to note: I was seeing issues where the error message was not dynamically changed on language changes from other devices. But I am able to see the same issue in main as well. So Will raise a separate bug on the same.

**Pull request merged**
